### ISSUE TITLE
[FEATURE] Import file lists from other buildsystems

### ISSUE THREAD

**Issue by LRN** [mesonbuild/meson#5201]
A project that does not use meson exclusively (for example, a project that is transitioning from autotools to meson) might want to avoid keeping separate source lists (one in `meson.build`, one in `Makefile.am`). Makefiles support separating parts of themselves via `include` keyword, and the included file could be a simple list of variable assignments with file names:
```make
FOO_SOURCES = \
  bar.c \
  baz.c \
  zool.c \

```
It would be trivial to have a meson function (`makefile_list()`, for example) that slurps this up and turns it into a dictionary strings that would look like `{'FOO_SOURCES': ['bar.c', 'baz.c', 'zool.c']}`. Barf at any hint of `$()` or `${}`, trim spaces before `=`, split the value after `=` by `' '` and filter out empty strings - and we're set.

This might also be expanded to other buildsystems that have some way of listing filenames in a separate files. Technically, this can be used for non-filename purposes, such as setting flags (`FOO_CFLAGS=-Wbar`), setting options (`FOO_USE_BAR=1`) - anything that can be expressed as a string and does not use any kind of variable expansion. That would be useful for projects that still make use of plain makefiles (`git`, i'm looking at you).

**Comment by Akaricchi** [comment]
FWIW, you can do this already with a simple python script to parse the makefile and print out the sources, and `run_command()` on the meson side to invoke it. Limited makefile parsing as a core feature seems a bit too specific to me.

**Comment by jpakkane** [comment]
> and run_command() on the meson side to invoke it

Yes, basically:

```meson
farr = run_command(...).stdout().strip().split()
```

> Technically, this can be used for non-filename purposes,

Never! We already have too many ways of defining things. We should have less of them, not more.

**Comment by dcbaker** [comment]
I wrote a module for doing this a long time ago, (just reading sources), but decided it just wasn't worth it, for a couple of reasons:
1) ultimately you want to minimize the number of build systems in your project, each one is a maintenance burden. Since meson is more capable than a make based system, you'll delete the make-based system.
2) android is moving away from a make syntax to a json-like purely declarative syntax. which makes having Makefile.sources to share with android builds a short term thing. (for android I'd really like to write a meson parser that sits on top of blueprint so that you can use meson natively in android)

**Comment by LRN** [comment]
> FWIW, you can do this already with a simple python script to parse the makefile and print out the sources, and `run_command()` on the meson side to invoke it

That seems to go against meson design. If that is how things are done, then you might as well add globbing.

**Comment by Akaricchi** [comment]
>That seems to go against meson design. If that is how things are done, then you might as well add globbing.

No, the way this particular thing is done is you put your source lists into meson.build. `run_command()` is a fine temporary workaround for a transitioning project, which is the use case you mentioned. No meson project should be parsing makefiles long-term, so it doesn't make any sense to have it as a core feature for that reason alone.

**Comment by bonzini** [comment]
The keyval module more or less does this (and it was created exactly to communicate between external scripts and Meson)

**Comment by eli-schwartz** [comment]
> FWIW, you can do this already with a simple python script to parse the makefile and print out the sources, and `run_command()` on the meson side to invoke it. Limited makefile parsing as a core feature seems a bit too specific to me.

With the filesystem module you instead do:

```meson
srcs = import('fs').read('sources.lst').strip().split('\n')
```

and fs.read() even handles making the file in question trigger reconfigures when it is edited.

More advanced substring parsing can be done if needed to support Makefile variable prefixes or other novel formats, if keyval isn't to one's taste.
